File: 1093-recover-a-tree-from-preorder-traversal/recover-a-tree-from-preorder-traversal.cs
public class Solution {
    public TreeNode RecoverFromPreorder(string traversal) {
        List<TreeNode> levels=new(); // keep track of the TreeNode at that depth
        for(int i=0; i<traversal.Length; ) {
            // read what depth this node is at
            int level=0;
            while(traversal[i]=='-') { i++; level++; }
            // read what value it should have
            int val=0;
            while(i<traversal.Length && traversal[i]!='-') val=val*10+traversal[i++]-'0';
            // create the node
            var node=new TreeNode(val);
            // add/set it to the levels list
            if(level<levels.Count) levels[level]=node; else levels.Add(node);
            // if it is not the root node, then connect this node to its parent
            if(level>0) {
                var parent=levels[level-1];
                if(parent.left!=null) { // if parent's .left is already set
                    parent.right=node;  // this is the .right child
                } else {
                    parent.left=node;   // otherwise it is the .left child
                }
            }
        }
        return levels[0]; // return the root TreeNode
    }
}

File: 1093-recover-a-tree-from-preorder-traversal/recover-a-tree-from-preorder-traversal.ts
function recoverFromPreorder(traversal: string): TreeNode | null {
    for (let count = 100; count > 0; count--) {
        traversal = traversal.replace(new RegExp("-".repeat(count), "g"), String.fromCharCode(count + 65));
    }

    function buildTree(parts: string, level: number): TreeNode | null {
        const splitParts = parts.split(String.fromCharCode(level + 65));
        const node = new TreeNode(parseInt(splitParts[0], 10));

        node.left = splitParts.length > 1 ? buildTree(splitParts[1], level + 1) : null;
        node.right = splitParts.length > 2 ? buildTree(splitParts[2], level + 1) : null;

        return node;
    }

    return buildTree(traversal, 1);
}